test/e2e/volume_provisioning.go

```diff
@@ -0,0 +1,197 @@
+
+			// Ugly hack: if we delete the AWS/GCE/OpenStack volume here, it will
+			// probably collide with destruction of the pods above - the pods
+			// still have the volume attached (kubelet is slow...) and deletion
+			// of attached volume is not allowed by AWS/GCE/OpenStack.
+			// Kubernetes *will* retry deletion several times in
+			// pvclaimbinder-sync-period.
+			// So, technically, this sleep is not needed. On the other hand,
+			// the sync perion is 10 minutes and we really don't want to wait
+			// 10 minutes here. There is no way how to see if kubelet is
+			// finished with cleaning volumes. A small sleep here actually
+			// speeds up the test!
+			// One minute should be enough to clean up the pods properly.
+			// Detaching e.g. a Cinder volume takes some time.
+			By("Sleeping to let kubelet destroy all pods")
+			time.Sleep(time.Minute)
+
+			By("deleting the claim")
+			expectNoError(c.PersistentVolumeClaims(ns).Delete(claim.Name))
+
+			// Wait for the PV to get deleted too.
+			expectNoError(waitForPersistentVolumeDeleted(c, pv.Name, 1*time.Second, 10*time.Minute))
+		})
+	})
+})
+
+func createClaim(ns string) *api.PersistentVolumeClaim {
+	return &api.PersistentVolumeClaim{
+		ObjectMeta: api.ObjectMeta{
+			GenerateName: "pvc-",
+			Namespace:    ns,
+			Annotations: map[string]string{
+				"volume.alpha.kubernetes.io/storage-class": "",
+			},
+		},
+		Spec: api.PersistentVolumeClaimSpec{
+			AccessModes: []api.PersistentVolumeAccessMode{
+				api.ReadWriteOnce,
+			},
+			Resources: api.ResourceRequirements{
+				Requests: api.ResourceList{
+					api.ResourceName(api.ResourceStorage): resource.MustParse(requestedSize),
+				},
+			},
+		},
+	}
+}
+
+// runInPodWithVolume runs a command in a pod with given claim mounted to /mnt directory.
+func runInPodWithVolume(c *client.Client, ns, claimName, command string) {
+	pod := &api.Pod{
+		TypeMeta: unversioned.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: api.ObjectMeta{
+			GenerateName: "pvc-volume-tester-",
+		},
+		Spec: api.PodSpec{
+			Containers: []api.Container{
+				{
+					Name:    "volume-tester",
+					Image:   "gcr.io/google_containers/busybox",
+					Command: []string{"/bin/sh"},
+					Args:    []string{"-c", command},
+					VolumeMounts: []api.VolumeMount{
+						{
+							Name:      "my-volume",
+							MountPath: "/mnt/test",
+						},
+					},
+				},
+			},
+			RestartPolicy: api.RestartPolicyNever,
+			Volumes: []api.Volume{
+				{
+					Name: "my-volume",
+					VolumeSource: api.VolumeSource{
+						PersistentVolumeClaim: &api.PersistentVolumeClaimVolumeSource{
+							ClaimName: claimName,
+							ReadOnly:  false,
+						},
+					},
+				},
+			},
+		},
+	}
+	pod, err := c.Pods(ns).Create(pod)
+	defer func() {
+		expectNoError(c.Pods(ns).Delete(pod.Name, nil))
+	}()
+	expectNoError(err, "Failed to create pod: %v", err)
+	expectNoError(waitForPodSuccessInNamespace(c, pod.Name, pod.Spec.Containers[0].Name, pod.Namespace))
```

Comment:
Might wanna use `waitForPodRunningInNamespaceSlow` here. Attach/detach can sometimes take several minutes to complete. See `test/e2e/pd.go` tests, for example.